File: 200-number-of-islands/200-number-of-islands.cs
public class Solution {
    public int NumIslands(char[][] grid) {
        int row = grid.Length;
        int col = grid[0].Length;
        int res = 0;
        
        for(int i = 0; i < row; i++)
        {
            for(int j = 0; j < col; j++)
                if(grid[i][j] == '1')
                {
                    backTrack(grid, i, j, row, col);
                    res++;
                }
        }
        return res;
    }
    
    public void backTrack(char[][] grid, int i, int j, int r, int c)
    {
        if(i < 0 || j < 0 || i >= r || j >= c || grid[i][j] == '0')
            return;
        grid[i][j] = '0';
        backTrack(grid, i + 1, j, r, c);
        backTrack(grid, i - 1, j, r, c);
        backTrack(grid, i, j + 1, r, c);
        backTrack(grid, i, j - 1, r, c);
    }
}

File: 200-number-of-islands/200-number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int row = grid.length;
        int col = grid[0].length;
        int res = 0;
        
        for(int i = 0; i < row; i++)
        {
            for(int j = 0; j < col; j++)
            {
                if(grid[i][j] == '1')
                {
                    backTrack(grid, i, j, row, col);
                    res++;
                }
            }
        }
        return res;
    }
    
    public void backTrack(char[][] grid, int i, int j, int row, int col)
    {
        if(i >= row || j >= col || i < 0 || j < 0 || grid[i][j] != '1')
            return;
        
        // mark as visited
        grid[i][j] = '0';
        backTrack(grid, i + 1, j, row, col);
        backTrack(grid, i - 1, j, row, col);
        backTrack(grid, i, j + 1, row, col);
        backTrack(grid, i, j - 1, row, col);
    }
}